### PR TITLE
Support pushing .sig files to the package upload endpoint

### PR DESCRIPTION
**Pull request by joelverhagen**
This adds support to the package push endpoint (`PUT /nuget`) to support uploaded .sig files at the same time as an uploaded  .nupkg. This is done by accepting a multipart request body. The first file in the multipart upload is assumed to be the package. This is existing behavior. Subsequent files in the multipart request body with the .sig file extensions are treated as detached signature files. Other files are ignored (this is also existing behavior).

This PR (as well as others code already in `package-signing` branch) is missing:

1. .sig validation for the associated .nupkg.
1. exposing signature files as JSON string property (design change).
1. reading the ID and version out of the PEM data (desgin change).

I also owe @dtivel a description of this endpoint as well as the signature CRUD endpoints for the design document.

### CODE REVIEW

File: src/NuGet.Server.Core/Infrastructure/TemporaryFileCollection.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    _logger.Log(
+                        LogLevel.Error,
+                        "Failed to delete temporary file {0}. Exception: {1}",
+                        path,
+                        exception);
+                }
+            }
+        }
+
+        public string GetPath()
```
Comment (dtivel) [inline src/NuGet.Server.Core/Infrastructure/TemporaryFileCollection.cs]: `GetNewFilePath(...)` would be a better name for 2 reasons:
* `GetPath(...)` never returns the same path.
* `GetPath(...)` always returns a path which is treated as a file path, which is also implied by the class name.

File: src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        /// <summary>
+        /// A dictionary where the key is the signature file name (user-provided) and the value is the temporary file
+        /// path (server-generated).
+        /// </summary>
+        public IReadOnlyDictionary<string, string> SignatureFilePaths { get; }
+
+        public static async Task<UploadedFiles> UploadAsync(
+            HttpRequestMessage request,
+            TemporaryFileCollection temp,
```
Comment (dtivel) [inline src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs]: Globally, `temporaryFiles` or `files` is a more self-explanatory name than `temp` when referring to a `TemporaryFileCollection` instance.

File: src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            var multipart = await request.Content.ReadAsMultipartAsync();
+
+            if (multipart.Contents.Count == 0)
+            {
+                return new UploadedFiles(UploadedFilesStatus.NoPackageFile);
+            }
+
+            // Upload the package.
+            var packagePath = await CopyFileAsync(multipart.Contents[0], temp);
```
Comment (dtivel) [inline src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs]: Should you check that the first element is a package, at least with a file extension check, or would that be a breaking change?

Comment (joelverhagen) [inline src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs]: The file extension check would be a breaking change. We later pass the first file to a `ZipPackage` constructor which does all the necessary checks to verify that it is indeed a package.

File: src/NuGet.Server.Core/Infrastructure/TemporaryFileCollection.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/TemporaryFileCollection.cs]: Fixed.

File: src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Server.V2/Infrastructure/UploadedFiles.cs]: Fixed.

**Pull request merged**